test_unstructured_ingest/unit/test_utils.py

```diff
@@ -0,0 +1,101 @@
+import json
+import typing as t
+from dataclasses import dataclass, field
+
+from unstructured.ingest.cli.utils import extract_config
+from unstructured.ingest.interfaces import BaseConfig
+
+
+@dataclass
+class A(BaseConfig):
+    a: str
+
+
+@dataclass
+class B(BaseConfig):
+    a: A
+    b: int
+
+
+flat_data = {"a": "test", "b": 4, "c": True}
+
+
+def test_extract_config_concrete():
+    @dataclass
+    class C(BaseConfig):
+        b: B
+        c: bool
+
+    c = extract_config(flat_data=flat_data, config=C)
+    expected_result = {"b": {"a": {"a": "test"}, "b": 4}, "c": True}
+    assert c.to_json(sort_keys=True) == json.dumps(expected_result, sort_keys=True)
+
+
+def test_extract_config_optional():
+    @dataclass
+    class C(BaseConfig):
+        c: bool
+        b: t.Optional[B] = None
+
+    c = extract_config(flat_data=flat_data, config=C)
+    expected_result = {"b": {"a": {"a": "test"}, "b": 4}, "c": True}
+    assert c.to_json(sort_keys=True) == json.dumps(expected_result, sort_keys=True)
+
+
+def test_extract_config_union():
+    @dataclass
+    class C(BaseConfig):
+        c: bool
+        b: t.Optional[t.Union[B, int]] = None
+
+    c = extract_config(flat_data=flat_data, config=C)
+    expected_result = {"b": 4, "c": True}
```

Comment:
Why is extracting `b` as an `int` favored in this test? 

When I checked the line below, I thought b as a dataclass (B) would be favored over b as an int for Unions; but then why the expected result is `{"b": 4, "c": True}` here?
https://github.com/Unstructured-IO/unstructured/blob/de5224f0c24c941c487fdbd28be5901a8123d0fa/unstructured/ingest/cli/utils.py#L78

Reply:
it only favors b as a dataclass if it doesn't already exist in the dictionary. It always defaults to the base class first if possible because that introduces less complex logic such as paring a new dataclass.

Reply:
Sounds good 👍